### PR TITLE
machine: introduce io.systemd.Machine.Open method

### PR DESCRIPTION
**Pull request by ikruglov**
This PR introduces io.systemd.Machine.Open method which combines three DBus alternatives:
- OpenMachinePTY
- OpenMachineLogin
- OpenMachineShell

The PR contains basic tests.

### CODE REVIEW

**Comment by yuwata** [comment]
Also, I'd like to merge #34882 at first. Currently the corresponding CI is unstable.

**Comment by ikruglov** [comment]
@bluca what are the changes?

**Comment by bluca** [comment]
> @bluca what are the changes?

The PR @yuwata was waiting for was merged, so just hit the "rebase" button to get those changes from latest main in, so we can see how the CI behaves

File: src/shared/varlink-io.systemd.Machine.c [hunk 1]
```diff
@@ -95,12 +95,41 @@ static SD_VARLINK_DEFINE_METHOD_FULL(
...
+                SD_VARLINK_FIELD_COMMENT("This mode allocates a pseudo TTY in the container and ensures that a getty login prompt of the container is running on the other end. It returns the file descriptor of the PTY and the PTY path. This is useful for acquiring a pty with a login prompt from the container."),
+                SD_VARLINK_DEFINE_ENUM_VALUE(login),
+                SD_VARLINK_FIELD_COMMENT("This mode allocates a pseudo TTY in the container, as the specified user, and invokes the executable at the specified path with a list of arguments (starting from argv[0]) and an environment block. It then returns the file descriptor of the PTY and the PTY path."),
+                SD_VARLINK_DEFINE_ENUM_VALUE(shell));
+
+static SD_VARLINK_DEFINE_METHOD(
+                Open,
+                VARLINK_DEFINE_MACHINE_LOOKUP_AND_POLKIT_INPUT_FIELDS,
+                SD_VARLINK_FIELD_COMMENT("There are three possible values: 'tty', 'login', and 'shell'. Please see description for each of the modes. Default value is 'tty'."),
+                SD_VARLINK_DEFINE_INPUT_BY_TYPE(mode, MachineOpenMode, SD_VARLINK_NULLABLE),
```
Comment (poettering) [inline src/shared/varlink-io.systemd.Machine.c]: "tty" shounds like a strange default, it's not that useful after all. Maybe not pick defaults here, but make this mandatory, i.e. drop the nullable?

File: src/machine/machine-varlink.c [hunk 2]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                                        "command_line", command_line);
+                default:
+                        assert_not_reached();
+        }
+}
+
+int vl_method_open(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dispatch_table[] = {
+                VARLINK_DISPATCH_MACHINE_LOOKUP_FIELDS(MachineOpenParameters),
+                { "mode", SD_JSON_VARIANT_STRING, json_dispatch_machine_open_mode, offsetof(MachineOpenParameters, mode), 0 },
```
Comment (poettering) [inline src/machine/machine-varlink.c]: as discussed below, I'd make this field mandatory

File: src/machine/machine-varlink.c [hunk 3]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                { "user", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, user), 0 },
+                { "path", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, path), 0 },
```
Comment (poettering) [inline src/machine/machine-varlink.c]: use json_dispatch_path

File: src/machine/machine-varlink.c [hunk 4]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                default:
+                        assert_not_reached();
+        }
+}
+
+int vl_method_open(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dispatch_table[] = {
+                VARLINK_DISPATCH_MACHINE_LOOKUP_FIELDS(MachineOpenParameters),
+                { "mode", SD_JSON_VARIANT_STRING, json_dispatch_machine_open_mode, offsetof(MachineOpenParameters, mode), 0 },
+                { "user", SD_JSON_VARIANT_STRING, sd_json_dispatch_const_string,   offsetof(MachineOpenParameters, user), 0 },
```
Comment (poettering) [inline src/machine/machine-varlink.c]: json_dispatch_const_user_group_name()

File: src/machine/machine-varlink.c [hunk 5]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+        if (r != 0)
+                return r;
+
+        if (p.mode == MACHINE_OPEN_MODE_SHELL) {
+                if (!isempty(p.user) && !valid_user_group_name(p.user, VALID_USER_RELAX))
+                        return sd_varlink_error_invalid_parameter_name(link, "user");
+                if (!isempty(p.path) && !path_is_absolute(p.path))
+                        return sd_varlink_error_invalid_parameter_name(link, "path");
+                // TODO(ikruglov): args validation?
+                if (!strv_isempty(p.env) && !strv_env_is_valid(p.env))
```
Comment (poettering) [inline src/machine/machine-varlink.c]: i sense that validating env blocks is so generic, that we should have json_dispatch_strv_env() for this too. we'll need it once we add varlink support for service activation in pid 1

Comment (poettering) [reply src/machine/machine-varlink.c]: heck, we already have such a function in src/shared/user-record.c: json_dispatch_environment()

can you move that to src/libsystemd/sd-json/json-util.[ch] and maybe rename it json_dispatch_strv_env() since we apparently otherwise call these strv with env assignments strv_env

File: src/machine/machine-varlink.c [hunk 6]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
+        if (r != 0)
+                return r;
+
+        if (p.mode == MACHINE_OPEN_MODE_SHELL) {
+                if (!isempty(p.user) && !valid_user_group_name(p.user, VALID_USER_RELAX))
```
Comment (poettering) [inline src/machine/machine-varlink.c]: i don't think we should use isempty() here. isempty() is intended for cases when the empty string and unset shall be treated the same. We use that often in D-BUS IPC code since D-Bus has no "nullable" concept, and you cannot emit fields, hence we often use the empty string as "niche" value for "unset". But in a json/varlink world that's unnecessary, we have a nullable concept. hence i think we should refuse empty usernames, but allow unset usernames.

tldr. change `!isempty(p.user)` to just `p.user`

File: src/machine/machine-varlink.c [hunk 7]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                        return sd_varlink_error_invalid_parameter_name(link, "user");
+                if (!isempty(p.path) && !path_is_absolute(p.path))
+                        return sd_varlink_error_invalid_parameter_name(link, "path");
+                // TODO(ikruglov): args validation?
+                if (!strv_isempty(p.env) && !strv_env_is_valid(p.env))
+                        return sd_varlink_error_invalid_parameter_name(link, "env");
+
+                user = isempty(p.user) ? "root" : p.user;
+                path = isempty(p.path) ? machine_default_shell_path() : p.path;
+                args = isempty(p.path) ? machine_default_shell_args(user) : strv_isempty(p.args) ? strv_new(path) : TAKE_PTR(p.args);
```
Comment (poettering) [inline src/machine/machine-varlink.c]: as above, let's not use isempty() for these

File: src/machine/machine-varlink.c [hunk 8]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                                return log_debug_errno(r, "Failed to start getty for machine '%s': %m", machine->name);
+
+                        break;
+
+                case MACHINE_OPEN_MODE_SHELL: {
+                        _cleanup_close_ int slave = -EBADF;
+
+                        slave = machine_open_terminal(machine, pty_name, O_RDWR|O_NOCTTY|O_CLOEXEC);
+                        if (slave < 0)
+                                return log_debug_errno(slave, "Failed to open terminal: %m");
```
Comment (poettering) [inline src/machine/machine-varlink.c]: maybe fow new code just avoid the "slave" term, to avoid fuss about it. just replace "slave" with "pty_fd" or so here, it's equally descriptive.

Comment (poettering) [reply src/machine/machine-varlink.c]: and maybe renae "master" to "ptmx_fd"...

File: src/machine/machine-varlink.c [hunk 9]
```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+                        break;
+                }
+
+                default:
+                        break;
+        }
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("ptyFileDescriptor", master),
```
Comment (poettering) [inline src/machine/machine-varlink.c]: to pass fds over varlink use sd_varlink_push_fd() and store the index integer it returns in the json.

File: src/shared/varlink-io.systemd.Machine.c [hunk 10]
```diff
@@ -131,4 +164,8 @@ SD_VARLINK_DEFINE_INTERFACE(
                 SD_VARLINK_SYMBOL_COMMENT("Machine uses a complex UID/GID mapping, cannot determine shift"),
                 &vl_error_NoUIDShift,
                 SD_VARLINK_SYMBOL_COMMENT("Requested information is not available"),
-                &vl_error_NotAvailable);
+                &vl_error_NotAvailable,
+                SD_VARLINK_SYMBOL_COMMENT("Requested operation is not supported"),
+                &vl_error_NotSupported,
+                SD_VARLINK_SYMBOL_COMMENT("There is no system bus in container"),
+                &vl_error_NoSDBus);
```
Comment (poettering) [inline src/shared/varlink-io.systemd.Machine.c]: hmm, we will sooner or later probably switch from D-Bus to varlink for this too, or try both, hence the error should be generic, and not encode D-Bus in its name.

maybe just "NoIPC"?

File: test/units/TEST-13-NSPAWN.machined.sh [hunk 11]
```diff
@@ -367,6 +367,17 @@ varlinkctl --more call /run/systemd/machine/io.systemd.Machine io.systemd.Machin
 (! varlinkctl --more call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.List '{"name": ".host"}' | grep 'acquireUIDShift')
 varlinkctl --more call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.List '{"name": ".host", "acquireMetadata": "yes"}' | grep 'UIDShift'
 
+# test io.systemd.Machine.Open
+varlinkctl call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.Open '{"name": ".host", "mode": "tty"}'
+varlinkctl call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.Open '{"name": ".host", "mode": "login"}'
+varlinkctl call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.Open '{"name": ".host", "mode": "shell"}'
+
+# TODO(ikruglov): how to really test other modes?
```
Comment (poettering) [inline test/units/TEST-13-NSPAWN.machined.sh]: hmm, that's a good question, i am not sure how we could make varlinkctl spit out fds in any reasonable form that the shell could use them... shells always work the other way round: the shell allocates an fd, and pass it to other processes, but the shall cannot accept fds from other processes.

Comment (poettering) [reply test/units/TEST-13-NSPAWN.machined.sh]: ah, here's an idea: maybe we should introduce a new "--exec=" switch or so for varlinkctl. You pass a cmdline of your choice, and we'll chainload that cmdline after a successful method call, and pass the method reply body as JSON input as stdin (`acquire_data_fd()` is great for that) and then pass any fds we received via fd 3 and bove.

we could even set $LISTEN_FDS env var to the number of fds received, so that the invoked process could use sd_listen_fds() to see what it got passed....

This would then mean you could do:

```c
varlinkctl call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.Open '{"name": ".host", "mode": "shell"}' --exec='bash -e "echo foo >&3 ; read x <&3 ; test $x == foo"'
```

or something like that. you get the gist.

File: src/machine/machine-varlink.c (hunk 9 above)
Comment (ikruglov) [reply src/machine/machine-varlink.c]: what do I do if `allow_fd_passing_output` is not enabled? just pass only name?

File: test/units/TEST-13-NSPAWN.machined.sh (hunk 11 above)
Comment (ikruglov) [reply test/units/TEST-13-NSPAWN.machined.sh]: ok make sense. I'd put it in a separate PR though. Mind merging without it and I'll follow up?

File: src/machine/machine-varlink.c (hunk 9 above)
Comment (ikruglov) [reply src/machine/machine-varlink.c]: I think I found answer: I need to enable it via `sd_varlink_set_allow_fd_passing_output()`

File: src/machine/machine.c [hunk 12]
```diff
@@ -702,6 +703,247 @@ int machine_open_terminal(Machine *m, const char *path, int mode) {
...
+        /* Auxiliary units */
+        r = sd_bus_message_append(tm, "a(sa(sv))", 0);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_call(container_bus, tm, 0, error, NULL);
+        if (r < 0)
+                return r;
+
+        pty_fd = safe_close(pty_fd);
```
Comment (poettering) [inline src/machine/machine.c]: unnecessary, you alradead have decalre `pty_fd` with `_cleanup_close_` at the top

File: src/machine/machine-varlink.c [hunk 13]
```diff
@@ -375,3 +378,195 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
...
+        ptmx_fd_idx = sd_varlink_push_fd(link, ptmx_fd);
+        /* no need to handle -EPERM because we do sd_varlink_set_allow_fd_passing_output() above */
+        if (ptmx_fd_idx < 0)
+                return log_debug_errno(ptmx_fd_idx, "Failed to push file descriptor over varlink: %m");
+
+        TAKE_FD(ptmx_fd);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("ptyFileDescriptor", ptmx_fd_idx),
```
Comment (poettering) [inline src/machine/machine-varlink.c]: doesn't matter, but strictly speaking ptmx_fd_idx is a signed entity, and you pass it here via SD_JSON_BUILD_PAIR_UNSIGNED(). Which is a bit weird. Not a real issue, because you verified above that this is not negative, but I still would make this SD_JSON_BUILD_PAIR_INTEGER() to avoid any raised eyebrows.

**Comment by poettering** [comment]
lgtm

**Comment by bluca** [comment]
```
../src/machine/machine-dbus.c:345:46: error: unused variable 'slave' [-Werror,-Wunused-variable]
        _cleanup_close_ int master = -EBADF, slave = -EBADF;
```

**Comment by ikruglov** [comment]
@bluca I've removed unused variable.

**Comment by bluca** [comment]
TEST-13-NSPAWN test failure looks new https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux/18714/console

**Comment by poettering** [comment]
needs rebase

**Comment by poettering** [comment]
CI appears to be unhappy still

**Comment by ikruglov** [comment]
@poettering @bluca CI is happy now. The "fix" is weird: https://github.com/systemd/systemd/pull/34867/commits/a6923441cd8f08edf7f0b1297398dc9101a66353.

I suspect that this is due to some problem in CI setup. It picked warning logs (see below) as a fatal error. The warning logs is due to passing invalid params as part of testing. 

```
TEST-13-NSPAWN: [   44.364317] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
[   44.400806] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': Wrong medium type
[   44.434151] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
[   44.467366] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': Wrong medium type
```

**Comment by ikruglov** [comment]
> I don't think log matters, the exit code must be changing due to changes in this PR. Sorry but this needs to be fixed, just disabling tests is not enough

```
[  129.392253] TEST-13-NSPAWN.sh[5038]: PASSED TESTS:   5:
[  129.392253] TEST-13-NSPAWN.sh[5038]: ------------------
[  129.392253] TEST-13-NSPAWN.sh[5038]: /usr/lib/systemd/tests/testdata/units/TEST-13-NSPAWN.importctl.sh
[  129.392253] TEST-13-NSPAWN.sh[5038]: /usr/lib/systemd/tests/testdata/units/TEST-13-NSPAWN.machined.sh
[  129.392253] TEST-13-NSPAWN.sh[5038]: /usr/lib/systemd/tests/testdata/units/TEST-13-NSPAWN.nspawn-oci.sh
[  129.392253] TEST-13-NSPAWN.sh[5038]: /usr/lib/systemd/tests/testdata/units/TEST-13-NSPAWN.nspawn.sh
[  129.392253] TEST-13-NSPAWN.sh[5038]: /usr/lib/systemd/tests/testdata/units/TEST-13-NSPAWN.nss-mymachines.sh
[  129.392940] TEST-13-NSPAWN.sh[482]: + touch /testok
[  129.437467] systemd-executor[5041]: SELinux enabled state cached to: disabled
[  129.439413] (setup.sh)[5041]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
[  129.493805] (setup.sh)[5041]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
[  129.495666] (setup.sh)[5041]: Successfully forked off '(sd-mkuserns)' as PID 5042.
[  129.497929] (setup.sh)[5041]: ID-mapping is supported for exec directory /var/lib/TEST-13-NSPAWN
[  129.498835] (setup.sh)[5041]: TEST-13-NSPAWN.service: Executing: /usr/lib/systemd/tests/integration-test-setup.sh finalize
[  129.507459] integration-test-setup.sh[5041]: + set -o pipefail
[  129.507584] integration-test-setup.sh[5041]: + case "$1" in
[  129.508157] integration-test-setup.sh[5043]: ++ systemctl list-jobs
[  129.571654] integration-test-setup.sh[5041]: + [[ JOB UNIT                   TYPE  STATE
[  129.571654] integration-test-setup.sh[5041]: 112 end.service            start waiting
[  129.571654] integration-test-setup.sh[5041]: 107 TEST-13-NSPAWN.service start running
[  129.571654] integration-test-setup.sh[5041]: 1   testsuite.target       start waiting
[  129.571654] integration-test-setup.sh[5041]: 3 jobs listed. =~ reboot.target|kexec.target|soft-reboot.target ]]
[  129.571654] integration-test-setup.sh[5041]: + rm -f /var/lib/TEST-13-NSPAWN/inprogress
[  132.123257] systemd[1]: end.service: Failed with result 'exit-code'.
[  132.124091] systemd[1]: Failed to start end.service.
Spawning getter /systemd-meson-build/journalctl -o export -D /var/tmp/systemd-test-TEST-13-NSPAWN_2/root/var/log/journal...
Creating journal file /var/tmp/systemd-test-TEST-13-NSPAWN_2/system.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
An error was encountered while opening journal file or directory /var/tmp/systemd-test-TEST-13-NSPAWN_2/root/var/log/journal/1c414e86d93a4fe8a138092c203c73ae, ignoring file: No such file or directory
/var/tmp/systemd-test-TEST-13-NSPAWN_2/system.journal: Journal header limits reached or header out-of-date, rotating
Creating journal file /var/tmp/systemd-test-TEST-13-NSPAWN_2/system.journal on a btrfs file system, and copy-on-write is enabled. This is likely to slow down journal access substantially, please consider turning off the copy-on-write file attribute on the journal directory, using chattr +C.
Vacuuming done, freed 0B of archived journals from /var/tmp/systemd-test-TEST-13-NSPAWN_2.
Finishing after writing 168026 entries
TEST-13-NSPAWN: [   44.364317] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
[   44.400806] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': Wrong medium type
[   44.434151] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
[   44.467366] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': Wrong medium type
-rw-r----- 1 root root 33554432 Nov  4 12:56 /var/tmp/systemd-test-TEST-13-NSPAWN_2/system.journal
qemu test failed
```

Here is the relevant log. As you can see all 5 parts of TEST-13-NSPAWN passed.

**Comment by poettering** [comment]
```
TEST-13-NSPAWN: [   44.364317] H systemd-machined[646]: varlink-15-15: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
```

Seems the IDL data of the Open() call doesn't match what you are passing to it.

**Comment by ikruglov** [comment]
To confirm my statements:

```
$ sudo mkosi boot
....

[root@default ~]# systemctl service-log-level systemd-machined debug
[root@default ~]# journalctl -q -o short-monotonic --grep "didn't pass validation" | grep -v "test-varlink-idl"
[root@default ~]# varlinkctl call /run/systemd/machine/io.systemd.Machine io.systemd.Machine.Open '{"name": ".host"}'
Method call io.systemd.Machine.Open() failed: org.varlink.service.InvalidParameter
{
        "parameter" : "mode"
}
[root@default ~]# journalctl -q -o short-monotonic --grep "didn't pass validation" | grep -v "test-varlink-idl"
[11568.867879] default systemd-machined[146]: varlink-14-14: Parameters for method io.systemd.Machine.Open() didn't pass validation on field 'mode': No anode
```

**Comment by ikruglov** [comment]
@poettering @bluca I see nothing wrong with this PR. If you want me to change/fix CI in a particular way, please let me know how. I'd happy to.

**Comment by ikruglov** [comment]
> Please fix the tests instead of commenting them out
@bluca can you please look into I wrote above. I don't see a way of fixing the tests. If you do please let me know. The problem is in CI setup.

**Comment by bluca** [comment]
If the error messages are expected, simply redirect those outputs to dev/null or something like th at

**Comment by ikruglov** [comment]
> If the error messages are expected, simply redirect those outputs to dev/null or something like th at

Sorry, but you're clearly not reading what I wrote. The errors messages are in journal. The end.sh [picks them up](https://github.com/systemd/systemd/pull/34867#issuecomment-2456866824) at the end of the test.

**Comment by bluca** [comment]
That's because the output of the script is captured in the journal

**Comment by ikruglov** [comment]
> That's because the output of the script is captured in the journal

No, it's not. Please check this https://github.com/systemd/systemd/pull/34867#issuecomment-2456886892

**Comment by bluca** [comment]
Then try to redirect it/change log level/change log target of whatever the source is

**Comment by poettering** [comment]
then maybe set an env var that skips that IDL invalid check if you validate cases like that?

**Comment by ikruglov** [comment]
I've pushed this commit  as a workaround https://github.com/systemd/systemd/pull/34867/commits/a08d805c9e99de83459d4c21419294ec13453fae. Let me know if you're happy with it

**Comment by bluca** [comment]
Please find a way to suppress the output as mentioned earlier instead, otherwise it's impossible to mix good and bad test cases

File: test/units/TEST-13-NSPAWN.machined.sh (hunk 11 above)
Comment (ikruglov) [reply test/units/TEST-13-NSPAWN.machined.sh]: @poettering, I've been thinking about your suggestion. Is there a real-life case where this --exec option would be useful? I mean: adding something just to run a test is too much IMO. There should be a simpler option if it's only for testing.
